### PR TITLE
Huge refactoring of packr to address several issues

### PR DESCRIPTION
**Pull request by karlsabo**
1. Converted packr to use Gradle
1. Include DrOpt source for easier building
1. Include sajson.h for easier building
   * <https://github.com/chadaustin/sajson/tree/791799ad90f7179f132ea2f53b90ef98f1d399a2>
   * From inspecting the old fips config at <https://github.com/code-disaster/fips-sajson>
1. Support tar.gz files
1. Support macOS signing and notarization of the executable
1. Load the msvcr*.dll that the JRE ships with instead of always trying to load msvcr100.dll
1. Removed linux32 platform
   * Linux x86 is no longer built because it's impossible to find a survey that shows anyone running x86 Linux.
1. Remove macOS x86 (32-bit) support
   * macOS x86 is no longer built because it requires and older version of Xcode and Apple makes it difficult to install on newer versions of macOS
1. Remove windows32 platform
   * Windows x86 is no longer built because the Adopt OpenJDK 8u242 and 8u252 have crash failures.
1. Added support for Java 11 and 14
   * Including jlink created JREs
   * Note: Packr cannot take a module and load it into the classpath of the created JVM, a Jar is still needed. It should be possible to create a jlink JRE from your custom module and have that work with packr but it's untested.
1. Added `useZgcIfSupportedOs` flag making it easier to use the Z garbage collector when bundling Java 14+.
   * The launcher executable will detect if the running operating system supports the Z garbage collector and use it.
2. Added [Google Test](https://github.com/google/googletest) C++ test framework to the [PackrLauncher code](PackrLauncher/src/test/cpp).
4. Updated to C++14 as the minimum supported C++ version.
1. Added support for unicode directories on Windows.
   * This resolves issues where PackrLauncher is running from a directory with international characters in it.
2. Fixed `--console` on Windows.
   * When running from a Windows Explorer shortcut, a new console is popped up and if `--verbose` was also specified then all debug output shows up in the console window.
3. If the PackrLauncher parent process has a console, PackrLauncher attaches to it.
   * This allows debug output when PackrLauncher is launched from a command prompt window, without the need for passing `--console`.
1. Resolves an issue for newer JVMs that rely on vcruntime140.dll (The Visual C++ 2017 Redistributable).
   * If loading the jvm.dll fails on Windows, then PackrLauncher searches for a vcruntime*.dll file in "jre/bin" and loads that library and attempts to load the jvm.dll again. This resolves an issue where the jvm.dll can't be loaded on Windows systems that don't have the Visual C++ 2017 Redistributable installed.
1. Support macOS versions down to 10.10 Yosemite
   * Added compiler flag `-mmacosx-version-min=10.10`
1. Fixed an issue with uncaught exception handlers not being called for the main thread.
   * dispatchUncaughtException is called on the main thread if an exception is detected after calling the main method.
1. Compile with `-no-pie` on Linux to work around a Nautilus bug.
   * <https://stackoverflow.com/questions/41398444/gcc-creates-mime-type-application-x-sharedlib-instead-of-application-x-applicati>
   * <https://stackoverflow.com/questions/34519521/why-does-gcc-create-a-shared-object-instead-of-an-executable-binary-according-to?noredirect=1&lq=1>
1. Added compile flags `/Os`, `/Gw`, `/Gy` on Windows.
   * Combined with the new linker flags, this reduced the executable size.
1. Added `/opt:icf`, `/opt:ref` linker flags on Windows.
   * Combined with the new compiler flags, this reduced the executable size.

### CODE REVIEW

**Comment by code-disaster** [comment]
Ok, that's been quite a bunch.

I hope you can forgive me for not stepping through each and every commit, but I had a fair look at the net result, and ... well, overall it looks fine from over here.

Since you pretty much restructured the whole project, it isn't easy to spot each and every change, and I wouldn't have gone the same direction as you did (some splosh of modern C++ here and there, using Gradle/Kotlin for native builds - I don't have much experience with that), but I ran it locally and it seems to work, so who am I to complain? :)

**Comment by MrStahlfelge** [comment]
So the real question here is, are you ready and willing to maintain the project with your changes @karlsabo ?

**Comment by karlsabo** [comment]
Thanks for taking the time to review it @code-disaster. I've been really happy with Gradle for building native code and it makes it easy to pull in the artifacts into the Java Gradle projects so I get quick turnaround on the C++ changes. I'd like to do some more C++ cleanup, specifically convert to std::filesystem, not sure I'll get around to it since the code is working well.

@MrStahlfelge I plan on maintaining it for as long as I can, I use it in our old games that I'm going to continue to support as much as possible. It's also a fun self contained project. I have dreams of it being my Kotlin/Native test bed 😁 too, no more C++code 😱. But that's really ambitious and unlikely to happen.

I'm going to leave this open a few more days to give others a chance to look it over and let me know if there's anything that needs modification.

Thanks again for the feedback.

**Comment by tommyettinger** [comment]
Wow. This is a ton of work! Are you using it (or preparing to use it) on some released project? I'm curious if customizing the icon on Windows targets will use a similar approach as before -- editing the C++ source and rebuilding `packr.exe`. It would really help to know what the minimum (or just recommended) versions of any build tools involved are, for any steps that need more than just Java. I'm still on Windows 7 for now, so my build tools may not be able to be as up-to-date -- C++14 should be pretty easy though.

File: CHANGES [hunk 1]
```diff
@@ -1,12 +0,0 @@
-[2.1-SNAPSHOT]
```
Comment (tommyettinger) [inline CHANGES:1]: Why was this deleted? There wasn't much there, to be fair...

File: Packr/packr.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+         existingFilename.contains("linux") && existingFilename.contains("x86") -> {
+            "packr-linux"
+         }
+         existingFilename.contains("mac") -> {
+            "packr-mac"
+         }
+         existingFilename.contains("windows") && existingFilename.contains("x86-64") -> {
+            "packr-windows-x64.exe"
+         }
+         existingFilename.contains("windows") && existingFilename.contains("x86") -> {
```
Comment (tommyettinger) [inline Packr/packr.gradle.kts:141]: I thought x86 Windows wasn't supported... I guess it doesn't matter? This could actually be good to keep in if the OpenJDK x86 builds are fixed, though x86 Windows has a tiny market share now. Still bigger than MacOS, probably.

File: Packr/src/main/java/com/badlogicgames/packr/Packr.java [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+	 /**
+	  * Install application-side file filter to specify which (additional) files can be deleted during the removePlatformLibs phase.
+	  * <p>
+	  * This filter is checked first, before evaluating the "--removelibs" and "--libs" options.
+	  *
+	  * @param filter file filter for removing libraries
+	  *
+	  * @return true if file should be removed (deleted)
+	  */
+	 @SuppressWarnings("unused") public Packr setRemovePlatformLibsFileFilter (Predicate<File> filter) {
```
Comment (tommyettinger) [inline Packr/src/main/java/com/badlogicgames/packr/Packr.java:144]: Haha yes! Java 8 in a libGDX-related repository, praise be unto Karl!

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -0,0 +1,49 @@
+stages:
+  - publish_launcher
+  - build
+  - publish
+
+publisWindowsExecutable:
```
[the remark this answers is not in the document]
Comment (karlsabo) [reply .gitlab-ci.yml]: Thank you for the in-depth review! Fixed this in 555de687bf.

File: CHANGES (hunk 1 above)
Comment (karlsabo) [reply CHANGES:1]: It should have been renamed to changelog.md. I added them to the changelog file now 54c1b14.

Thanks.

**Comment by karlsabo** [comment]
> 
> 
> Wow. This is a ton of work! Are you using it (or preparing to use it) on some released project? I'm curious if customizing the icon on Windows targets will use a similar approach as before -- editing the C++ source and rebuilding `packr.exe`. It would really help to know what the minimum (or just recommended) versions of any build tools involved are, for any steps that need more than just Java. I'm still on Windows 7 for now, so my build tools may not be able to be as up-to-date -- C++14 should be pretty easy though.

@tommyettinger yeah, I'm using it for [Flamebreak](https://store.steampowered.com/app/399640/Flamebreak/). I haven't looked into the custom icon issue. C++14 on Windows should be as easy as installing Visual Studio 2017. I'm going to work on getting the code to compile and run using MinGW so there's an open source build path available on Windows.
